### PR TITLE
Do not throw errors from DistributedContext in runtime

### PR DESCRIPTION
**Pull request by lmolkova**
Next part of #295. OTel error policy to NOT throw in runtime and fallback to invalid/empty values.

### CODE REVIEW

File: src/OpenTelemetry.Api/DistributedContext/DistributedContext.cs [hunk 1]
```diff
@@ -116,10 +111,12 @@ public static DistributedContextCarrier Carrier
             {
                 if (value is null)
                 {
-                    throw new ArgumentNullException(nameof(value));
+                    OpenTelemetryApiEventSource.Log.InvalidArgument("set_Carrier", "value is null");
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry.Api/DistributedContext/DistributedContext.cs]: in this case let's assign no-op carrier to the Carrier field. Perhaps that is what user meant by setting it to `null`

File: src/OpenTelemetry.Api/DistributedContext/DistributedContextEntry.cs [hunk 2]
```diff
@@ -41,8 +41,26 @@ public DistributedContextEntry(string key, string value)
         /// <param name="metadata">Entry metadata.</param>
         public DistributedContextEntry(string key, string value, in EntryMetadata metadata)
         {
-            this.Key = key ?? throw new ArgumentNullException(nameof(key));
-            this.Value = value ?? throw new ArgumentNullException(nameof(value));
+            if (key == null)
+            {
+                this.Key = string.Empty;
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry.Api/DistributedContext/DistributedContextEntry.cs]: is dictionary fine with empty string keys? Maybe we can use `"(null)"` instead?

File: src/OpenTelemetry.Api/DistributedContext/DistributedContextEntry.cs [hunk 3]
```diff
@@ -41,8 +41,26 @@ public DistributedContextEntry(string key, string value)
...
+                OpenTelemetryApiEventSource.Log.InvalidArgument(nameof(DistributedContextEntry), "null key");
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry.Api/DistributedContext/DistributedContextEntry.cs]: maybe better to use nameof(key) and last string to be a reason (optional)

File: src/OpenTelemetry.Api/DistributedContext/DistributedContextEntry.cs (hunk 2 above)
Comment (lmolkova) [reply src/OpenTelemetry.Api/DistributedContext/DistributedContextEntry.cs]: DistributedcontextEntries are stored in the list
empty string is typical representation of null in c#, also we use empty string as fallbacks everywhere else. why here it should be different?

**Comment by SergeyKanzhelev** [comment]
sorry for another quick merge. Happy to retroactively address concerns

**Pull request merged**